docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,436 @@
+    - **InfrastructureRef [required]**
+        - Type: `corev1.TypedLocalObjectReference`
+        - Description: InfrastructureRef is a required reference to a Custom Resource Definition offered by an infrastructure provider.
+    - **Version [optional]**
+        - Type: `string`
+        - Description: Version defines the desired Kubernetes version. This field is meant to be optionally used by bootstrap providers.
+
+```go
+type MachineStatus struct
+```
+
+- **To remove**
+    - **ProviderStatus [optional]**
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderStatus details a Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `*MachineVersionInfo`
+        - Description: Versions specifies the current versions of software on the corresponding Node (if it exists). This is provided for a few reasons [...].
+- **To add**
+    - **Version [optional]**
+        - Type: `*string`
+        - Description: Version specifies the current version of Kubernetes running on the corresponding Node. This is meant to be a means of bubbling up status from the Node to the Machine. It is entirely optional, but useful for end-user UX if it’s present.
+    - **Bootstrap [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **Infrastructure [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
+```
+- **To add**
+    - **ConfigRef [optional]**
+        - Type: `*corev1.TypedLocalObjectReference`
+        - Description: ConfigRef is a reference to a bootstrap provider-specific resource that holds configuration details. The reference is optional to allow users/operators to specify Bootstrap.Data without the need of a controller.
+    - **Data [optional]**
+        - Type: `*string`
+        - Description: Data contains the bootstrap data, such as cloud-init details scripts. If nil, the Machine should remain in the Pending state.
+
+## States and transitions
+The Cluster API Machine Controller is responsible for managing the lifecycle of a Machine, including its state transitions. This is the only controller that should be writing updates to Machines.
+
+---
+### Pending
+```go
+// MachinePhasePending is the first state a Machine is assigned by
+// Cluster API Machine controller after being created.
+MachinePhasePending = MachinePhaseType("pending")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is empty string
+
+#### Expectations
+- When `Machine.Spec.Bootstrap.Data` is:
+    - `<nil>`, expect the field to be set by an external controller.
+    - `""` (empty string), expect the bootstrap step to be ignored.
```

Comment:
I'm not understanding why `Machine.Spec.Bootstrap.Data` strictly requires a value. Wouldn't this be up to the bootstrap provider to interpret what it means? This relates to https://github.com/kubernetes-sigs/cluster-api/pull/997/files#r293168090 too.

Reply:
While I completely agree with you, some folks have expressed the wish to disable the bootstrap step, which is allowed by setting an empty string for the field. The Machine controller would just see it not-nil and proceed with the state transitions as described below.

Reply:
Right, the idea is to provide some way for code looking at the Machine to determine if bootstrapping has been done, without relying on a phase/state field with values that might change. The presence or absence of data in the Bootstrap.Data field is a simple check.

Reply:
Re-reading this, I'm not sure we need to say that an empty string means the bootstrap step will be ignored. Perhaps the bootstrap data is stored somewhere else (a secret) or managed in some other way entirely that means it doesn't have to be stored here. An empty string could just as easily mean "bootstrapping is done" in those cases.

Reply:
This thread seems to be on the same topic as [the one above](https://github.com/kubernetes-sigs/cluster-api/pull/997#discussion_r293168090).

I agree with @dhellmann's description. However, in light of @davidewatson's comment, I think this proposal will use Conditions instead of Phases.

Let's consider using a condition to indicate that the bootstrap data is ready.  For example, to indicate that the bootstrap data is ready, add a condition to the Machine Status: type `BootstrapDataReady`, status `true`.

Reply:
Something has to know when to set `BootstrapDataReady` to true. For now, would you be amenable to the tri-state value of the field, with `nil` meaning not ready, and any string value, including the empty string, meaning ready? We can revisit in v1alpha3 if needed.

Reply:
@ncdc, I don't even think we need to consider it tri-state. If pointer is not nil then bootstrapping has been done.

Reply:
That's what I get for trying to comment while on the phone. You are correct - nil and not nil are the only 2 states that matter here.

Reply:
> Something has to know when to set `BootstrapDataReady` to true

Did I misunderstand something? [Figure 5](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/c6f48c5597e67e39933bfa2449c789791aac8018/docs/proposals/images/machine-states-preboot/Figure5.png) shows Machine controller copying over the data from the bootstrap CR. The Machine controller knows the data is ready by checking the Phase of the bootstrap CR. I'm proposing the Machine controller write a condition to the Machine status when it copies over the data.

Reply:
@dlipovetsky you are correct - yesterday was a looooong day for me 😄. We can do what you suggested, as well as have the machine controller set the status to true if it sees the bootstrap data is non nil (e.g. the user set it themselves).

Reply:
@ncdc NP :smile: I'm glad we're on the same page. I didn't say this earlier, but it occurred to me that being explicit (using Conditions) has another plus: it will help humans (probably unfamiliar with CAPI implementation details) understand the "progress" of some resource.